File: ui/communication/serial_writer.py
import struct
from .serial_reader import SerialReader

class SerialWriter():
    def __init__(self, serialPort):
        self.serialPort = serialPort

    def sendPacket(self, code, data=None):
        if data == None: data = struct.pack("< x")
        size = len(data)
        print("size is {}".format(size))

        packet_body = bytes()
        packet_body += struct.pack("< Bh", code, size)
        packet_body += data

        crc = 0x00
        for b in packet_body:
            crc = crc ^ b

        packet = bytes()
        packet += struct.pack("< BB", SerialReader.PACKET_HEADER1, SerialReader.PACKET_HEADER2)
        packet += packet_body
        packet += struct.pack("< B", crc)

        self.serialPort.write(packet)
        print(packet)

    #def request_gyro_acc(serialPort):
    #    while True:
    #        time.sleep(0.1)
    #        sendPacket(serialPort, 2)

File: js_ui2/app/serial_writer.js
import Utils from './utils'
import SerialCodes from './serial_codes'

class SerialWriter {
  constructor(send) {
    this.send = send
  }

  sendPacket(code, data = null, callback = null) {
    if (!data) { data = new ArrayBuffer(1) }
    const dataView = new Uint8Array(data)

    const size = 6 + data.byteLength
    let crc = 0

    let packetBuffer = new ArrayBuffer(size)
    let packetView = new Uint8Array(packetBuffer) 

    packetView[0] = SerialCodes.PACKET_HEADER1
    packetView[1] = SerialCodes.PACKET_HEADER2
    packetView[2] = code
    packetView[3] = Utils.lowByte(data.byteLength)
    packetView[4] = Utils.highByte(data.byteLength)

    crc = packetView[2] ^ packetView[3] ^ packetView[4]

    dataView.forEach(function(d, i) {
      packetView[i+5] = d
      crc ^= d
    })

    packetView[size-1] = crc

    // console.log('sending packet', packetView)
    this.send(packetBuffer)
  }
}

export default SerialWriter
